finality-aleph/src/sync/forest/mod.rs

```diff
@@ -0,0 +1,662 @@
+            HighestFinalized => (),
+            Unknown(_) | HopelessFork | BelowMinimal => return Err(Error::IncorrectParentState),
+        }
+        match self.get_mut(&id) {
+            Candidate(mut entry) => Ok(entry.get_mut().vertex.insert_body(parent_id.clone())),
+            _ => Err(Error::IncorrectVertexState),
+        }
+    }
+
+    /// Updates the provided justification, returns whether either finalization is now possible or
+    /// the vertex became a new top required.
+    pub fn update_justification(
+        &mut self,
+        justification: J,
+        holder: Option<I>,
+    ) -> Result<JustificationAddResult, Error> {
+        use JustificationAddResult::*;
+        let (id, parent_id) = self.process_header(justification.header())?;
+        self.update_header(justification.header(), None, false)?;
+        match self.get_mut(&id) {
+            VertexHandle::Candidate(mut entry) => {
+                match entry.get_mut().vertex.insert_justification(
+                    parent_id.clone(),
+                    justification,
+                    holder,
+                ) {
+                    Noop => Ok(Noop),
+                    Required => {
+                        self.top_required.insert(id.clone());
+                        self.set_required(&parent_id);
+                        Ok(Required)
+                    }
+                    Finalizable => {
+                        self.top_required.remove(&id);
+                        Ok(Finalizable)
+                    }
+                }
+            }
+            _ => Ok(Noop),
+        }
+    }
+
+    fn prune_level(&mut self, level: u32) {
+        let to_prune: Vec<_> = self
+            .vertices
+            .keys()
+            .filter(|k| k.number() <= level)
+            .cloned()
+            .collect();
+        for id in to_prune.into_iter() {
+            self.prune(&id);
+        }
+        self.compost_bin.retain(|k| k.number() > level);
+    }
+
+    /// Attempt to finalize one block, returns the correct justification if successful.
+    pub fn try_finalize(&mut self) -> Option<J> {
+        for child_id in self.root_children.clone().into_iter() {
+            if let Some(VertexWithChildren { vertex, children }) = self.vertices.remove(&child_id) {
+                match vertex.ready() {
+                    Ok(justification) => {
+                        self.root_id = child_id;
+                        self.root_children = children;
+                        self.prune_level(self.root_id.number());
+                        return Some(justification);
+                    }
+                    Err(vertex) => {
+                        self.vertices
+                            .insert(child_id, VertexWithChildren { vertex, children });
+                    }
+                }
+            }
+        }
+        None
+    }
+
+    /// How much interest we have for the block.
+    pub fn state(&mut self, id: &BlockIdFor<J>) -> Interest<I> {
```

Comment:
Yup, exactly. The idea is that we will request `TopRequired` more often than others (they still have to be sometimes requested, otherwise someone could make us only attempt requests for nonexisting blocks) and that `Required` blocks will get imported, while `Uninterested` not.